[JIT] Combine concat nodes where possible

mikeiovine opened the pull request:
Summary:
See the [related issue](https://github.com/pytorch/pytorch/issues/66654) for a description of the pattern that this pass optimizes.

Enabled the new pass in SR since we've seen cases where it is useful.

Test Plan: `buck test caffe2/test/cpp/jit:jit -- ConcatOpt`

Differential Revision: D31819491

eellison reviewed (commented):
Cool!!! didn't look too at the concat combine logic yet, but a couple things - also could we just hook this into peephole optimize, as we did with other graph optimizations ? This will make sure it's called accordingly throughout freezing, jit execution, static runtime etc

mikeiovine commented on torch/csrc/jit/passes/concat_opt.cpp:
```diff
@@ -497,5 +497,297 @@ void ExpandConcatAndEliminateRedundancy(const std::shared_ptr<Graph>& graph) {
+    }
+  }
+
+  bool eliminateTrivialNodes() {
```
I decided to abandon the trivial concat stuff. Aside from the aliasing issues, it can cause issues if you don't know the shape of the tensor statically. Specifically, you can have graphs like this:
```
def forward(x):
    return torch.cat([x], 3)
```
which will throw an exception for certain inputs:
```
x = torch.tensor(1.0)
forward(x)
# RuntimeError: zero-dimensional tensor (at position 0) cannot be concatenated
```
Since I haven't seen any models that could benefit from eliminating trivial concats (I implemented it more as a "nice to have" kind of thing), I think it's more trouble than it's worth. 

We conveniently ignore issues like this in a lot of static runtime passes. But since we're hooking this into `PeepholeOptimize`, we need to be more careful.

Commit:
[JIT] Combine concat nodes where possible (#67000)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/67000

See the [related issue](https://github.com/pytorch/pytorch/issues/66654) for context.

This new JIT optimization transforms patterns like this:
```
%inputs.1 : Tensor[] = prim::ListConstruct(%a, %b, %c)
%concat.1 : Tensor = aten::cat(%inputs, %dim)
%inputs.2 : Tensor[] = prim::ListConstruct(%x, %concat.1, %y)
%concat.2 : Tensor = aten::cat(%inputs.2, %dim)
```
into this:
```
%inputs.2 : Tensor[] = prim::ListConstruct(%x, %a, %b, %c, %y)
%concat.2 : Tensor = aten::cat(%inputs.2, %dim)
```
(it can do this for chains of `aten::cat` longer than 2 as well)

A few conditions have to hold:
1.  The `dim`s have to match.
2. `inputs.1` and `inputs.2` cannot be mutated

Test Plan: `buck test caffe2/test/cpp/jit:jit -- ConcatOpt`

Reviewed By: d1jang

Differential Revision: D31819491

fbshipit-source-id: be43fb2d0a9489a696c6966ff3db87e943ce7907
